qiskit_nature/algorithms/initial_points/hf_initial_point.py

```diff
@@ -0,0 +1,177 @@
+from qiskit_nature.properties.second_quantization.second_quantized_property import (
+    GroupedSecondQuantizedProperty,
+)
+
+from .initial_point import InitialPoint
+
+logger = logging.getLogger(__name__)
+
+
+class HFInitialPoint(InitialPoint):
+    r"""Compute the Hartree-Fock (HF) initial point.
+
+    A class that provides an initial point for the :class:`~qiskit.algorithms.VQE` parameter values
+    that will correspond to the HF initial point once the
+    :class:`~qiskit.circuit.library.HFInitialState` circuit is prepended to the
+    :class:`~qiskit.circuit.library.UCC` circuit.
+    The excitation list generated by the :class:`~qiskit.circuit.library.UCC` ansatz is obtained to
+    ensure that the shape of the initial point is appropriate.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._reference_energy: float = 0.0
+        self._excitation_list: list[tuple[tuple[int, ...], tuple[int, ...]]] | None = None
+        self._parameters: np.ndarray | None = None
+
+    def __len__(self) -> int:
+        """The length of the computed initial point array."""
+        return len(self.to_numpy_array())
+
+    @property
+    def grouped_property(self) -> GroupedSecondQuantizedProperty:
+        """The grouped property.
+
+        The grouped property is not required to compute the HF initial point.
+        """
+        return self._grouped_property
+
+    @grouped_property.setter
+    def grouped_property(self, grouped_property: GroupedSecondQuantizedProperty) -> None:
+        if not isinstance(grouped_property, GroupedSecondQuantizedProperty):
+            logger.warning(
+                "grouped_property is not of type GroupedSecondQuantizedProperty; "
+                "grouped_property and reference_energy will not be set."
+            )
+            return
+
+        electronic_energy: ElectronicEnergy | None = grouped_property.get_property(ElectronicEnergy)
+        if not isinstance(electronic_energy, ElectronicEnergy):
+            logger.warning(
+                "ElectronicEnergy not obtained from grouped_property; "
+                "grouped_property and reference_energy will not be set."
+            )
+            return
+
+        self._reference_energy = electronic_energy.reference_energy if not None else 0.0
+        self._grouped_property = grouped_property
+
+    @property
+    def ansatz(self) -> UCC:
+        """The UCC ansatz.
+
+        This is used to ensure that the :attr:`excitation_list` matches with the UCC ansatz that
+        will be used with the VQE algorithm.
+
+        Raises:
+            QiskitNatureError: If not set using a valid
+                :class:`~qiskit_nature.circuit.library.ansatzes.ucc.UCC` instance.
+        """
+        return self._ansatz
+
+    @ansatz.setter
+    def ansatz(self, ansatz: UCC) -> None:
+        if not isinstance(ansatz, UCC):
+            raise QiskitNatureError(f"Expected a UCC ansatz, but got type: {type(ansatz)}.")
```

Comment:
Yeah, I was conflicted about this. @mrossinek do you have a strong opinion? I could relax this instance checks in a number of places. It feels "unpythonic" to be this strict on the type at runtime, but it might catch the odd user input error earlier.

Reply:
I decided to remove the runtime type checking. I think the type hints should be sufficient for the user to debug their approach. The overhead/maintenance of type checking everything is a bad trend to set, I think. Happy to revert if there are objections.